test/view/set_difference.cpp

```diff
@@ -0,0 +1,317 @@
+        auto res1 = view::set_difference(s_finite, view::ints(-2, 10),
+                                         ordered_less(),
+                                         [](const S& x){ return x.val; },
+                                         ident()
+                                        );
+        using R1 = decltype(res1);
+        CONCEPT_ASSERT(Same<range_value_t<R1>, S>());
+        CONCEPT_ASSERT(Same<range_reference_t<R1>, S&>());
+        CONCEPT_ASSERT(Same<range_rvalue_reference_t<R1>, S&&> ());
+        ::check_equal(res1, {S{-20}, S{-10}, S{3}, S{20}});
+
+
+        auto res2 = view::set_difference(view::ints(-2, 10), s_finite,
+                                         ordered_less(),
+                                         ident(),
+                                         [](const S& x){ return x.val; }
+                                        );
+        using R2 = decltype(res2);
+        CONCEPT_ASSERT(Same<range_value_t<R2>, int>());
+        CONCEPT_ASSERT(Same<range_reference_t<R2>, int>());
+        CONCEPT_ASSERT(Same<range_rvalue_reference_t<R2>, int> ());
+        ::check_equal(res2, {-2, -1, 0, 2, 4, 5, 7, 9});
+    }
+    
+    
+    // move
+    {
+        auto v0 = to_<std::vector<MoveOnlyString>>({"a","b","b","c","x","x"});
+        auto v1 = to_<std::vector<MoveOnlyString>>({"b","x","y","z"});
+        auto res = view::set_difference(v0, v1, [](const MoveOnlyString& a, const MoveOnlyString& b){return a<b;});
+
+        std::vector<MoveOnlyString> expected;
+        move(res, back_inserter(expected));
+
+        ::check_equal(expected, {"a","b","c","x"});
+        ::check_equal(v1, {"b","x","y","z"});
+        //::check_equal(v0, {"","","b","","","x"});
+        
+        // FIXME: why is v0 not equal to {"","","b","","","x"}?
+        // the lazy view::set_difference mimics the greedy set_difference very closely!
+        
+        auto v0_greedy = to_<std::vector<MoveOnlyString>>({"a","b","b","c","x","x"});
+        auto v1_greedy = to_<std::vector<MoveOnlyString>>({"b","x","y","z"});
+        std::vector<MoveOnlyString> expected_greedy;
+        set_difference(v0_greedy, v1_greedy,
+                       move_into(back_inserter(expected_greedy)),
+                       [](const MoveOnlyString& a, const MoveOnlyString& b){return a<b;});
+        ::check_equal(expected_greedy, expected);
+        ::check_equal(v0_greedy, v0);
+        ::check_equal(v1_greedy, v1);
+
+ 
+        using R = decltype(res);
+
+        CONCEPT_ASSERT(Same<range_value_t<R>, MoveOnlyString>());
+        CONCEPT_ASSERT(Same<range_reference_t<R>, MoveOnlyString &>());
+        CONCEPT_ASSERT(Same<range_rvalue_reference_t<R>, MoveOnlyString &&>());
+    }
+    
+    // WARNING: set_difference between two infinite ranges can create infinite loops!
+    // {
+    //     auto empty_range = view::set_difference(view::ints, view::ints);
+    //     begin(empty_range); // infinite loop!
+    // }
```

Comment:
It can. That however does not avoid possibly infinite loops. There are no compile-time guarantees that the difference is not empty, in fact the issue is similar to starting a possibly never-ending `find_if()` on an infinite range.